cordapp-example/workflows-kotlin/src/main/kotlin/com/example/flow/PaybackFlow.kt

```diff
@@ -0,0 +1,160 @@
+ */
+object PaybackFlow {
+    @InitiatingFlow
+    @StartableByRPC
+    class Initiator(val inputLinearId: String, val paybackValue: Int) : FlowLogic<SignedTransaction>() {
+        companion object {
+            object CREATING_COMMAND : Step("Pull data from vault and create command.")
+            object GENERATING_TRANSACTION : Step("Generating transaction based on IOU state.")
+            object VERIFYING_TRANSACTION : Step("Verifying contract constraints.")
+            object SIGNING_TRANSACTION : Step("Signing transaction with our private key.")
+            object INITIALIZE_OTHER_PARTY_FLOW : Step("Send the state to the counterparty, and receive it back with their signature")
+
+            object GATHERING_SIGS : Step("Gathering the counterparty's signature.") {
+                override fun childProgressTracker() = CollectSignaturesFlow.tracker()
+            }
+
+            object PAYBACK_COMMUNICATION : Step("Sending payback value to lender.")
+
+            object FINALISING_TRANSACTION : Step("Obtaining notary signature and recording transaction.") {
+                override fun childProgressTracker() = FinalityFlow.tracker()
+            }
+
+            fun tracker() = ProgressTracker(
+                    CREATING_COMMAND,
+                    GENERATING_TRANSACTION,
+                    VERIFYING_TRANSACTION,
+                    SIGNING_TRANSACTION,
+                    INITIALIZE_OTHER_PARTY_FLOW,
+                    GATHERING_SIGS,
+                    PAYBACK_COMMUNICATION,
+                    FINALISING_TRANSACTION
+            )
+        }
+
+        override val progressTracker = tracker()
+
+        /**
+         * The flow logic is encapsulated within the call() method.
+         */
+        @Suspendable
+        override fun call(): SignedTransaction {
+            progressTracker.currentStep = CREATING_COMMAND
+            //val signedTransaction = serviceHub.validatedTransactions.getTransaction(loanTxId)
+            // val signedTransaction = transactions.find { it.id == transactionHash } ?: throw IllegalArgumentException("Unknown transaction hash.")
+
+            /* val criteria = QueryCriteria.VaultQueryCriteria(
+                    contractStateTypes = setOf(IOUState::class.java),
+                    status = Vault.StateStatus.UNCONSUMED,
+                    exactParticipants = listOf(ourIdentity, otherParty)
+            ) // */
+
+            val uuid: UUID = UUID.fromString(inputLinearId)
+            val queryCriteria: QueryCriteria = LinearStateQueryCriteria().withUuid(Collections.singletonList(uuid))
+
+            val results = serviceHub.vaultService.queryBy(IOUState::class.java, queryCriteria)
+            // val ourStateRef = StateRef(SecureHash.parse(loanTxId), 0)
+            // val (state, ref) = serviceHub.toStateAndRef<ContractState>(ourStateRef)
+            val inputStateAndRef = serviceHub.toStateAndRef<ContractState>(results.states[0].ref)
```

Comment:
No need to lose info to look for it afterwards. This will do:

```kotlin
val inputStateAndRef = results.states[0]
```
Even better, use `.single()` or `.first()`.
Mmh, actually, I would add a nice check that we got what we wanted:

```kotlin
if (results.states.size() != 1) throw FlowException("id not found or duplicate")
val inputStateAndRef =results.states.single()
```

Reply:
Plus, if you had done `.toStateAndRef<IOUState>`, you would not need an extra casting.